From c5702590080aa5d0e8d666d91861d64634065132 Mon Sep 17 00:00:00 2001
From: Aleksei Latyshev <alex_700_95@mail.ru>
Date: Sat, 21 May 2022 05:48:08 -0700
Subject: [PATCH] work around gcc-12 + CWG 903 in FOLLY_SAFE_CHECK (#1777)

Summary:
- it is workaround CWG 903 in gcc-12, which is considering `int{}`  as null pointer constant, and so can't choose between `safe_assert_msg_types_one(uint64_t)` and `safe_assert_msg_types_one(char const *)`
- see https://godbolt.org/z/fov5xPjEe

previously landed in {D36458136 (https://github.com/facebook/folly/commit/68ea0c2b3b9e041180b1bad4f96fe6775fe65744)}, which triggered some unknown bug and was reverted

closes https://github.com/facebook/folly/issues/1772

Pull Request resolved: https://github.com/facebook/folly/pull/1777

Test Plan:
Imported from GitHub, without a `Test Plan:` line.

  buck2 test folly/lang/test:safe_assert_test
  buck build fbsource//fbandroid/mode/server fbsource//fbandroid/mode/opt fbsource//fbandroid/apps/assistantplayground:assistantplayground_release

Reviewed By: Orvid

Differential Revision: D36552302

Pulled By: yfeldblum

fbshipit-source-id: 2bc00d2e6b893c2192c7071bbc859df73924c6f5
---
 folly/lang/SafeAssert.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/folly/lang/SafeAssert.h b/folly/lang/SafeAssert.h
index adf4d47e8fc..8ce9bc28022 100644
--- a/folly/lang/SafeAssert.h
+++ b/folly/lang/SafeAssert.h
@@ -130,7 +130,7 @@ FOLLY_INLINE_VARIABLE constexpr safe_assert_msg_types_one_fn
     safe_assert_msg_types_one{}; // a function object to prevent extensions
 
 template <typename... A>
-safe_assert_msg_type_s<decltype(safe_assert_msg_types_one(A{}))::value...>
+safe_assert_msg_type_s<decltype(safe_assert_msg_types_one((A)A{}))::value...>
 safe_assert_msg_types_seq_of(A...); // only used in unevaluated contexts
 
 template <typename>
